library/core/src/iter/traits/collect.rs

```diff
@@ -427,6 +429,22 @@ pub trait Extend<A> {
     fn extend_reserve(&mut self, additional: usize) {
         let _ = additional;
     }
+
+    /// Extends a collection with one element, without checking there is enough capacity for it.
+    ///
+    /// **Note:** For a collection to unsafely rely on this method's safety precondition (that is,
+    /// invoke UB if they are violated), it must implement `extend_reserve` correctly. In other words,
+    /// callers may assume that if they `extend_reserve`ed enough space they can call this method.
+    ///
+    /// # Safety
+    ///
+    /// This must only be called when we know the collection has enough capacity to contain the new item,
+    /// for example because we previously called `extend_reserve`.
+    #[unstable(feature = "extend_one_unchecked", issue = "none")]
+    #[doc(hidden)]
+    unsafe fn extend_one_unchecked(&mut self, item: A) {
+        self.extend_one(item);
+    }
```

Comment:
I don't think adding this method to a trait makes sense without intent for stabilization.

Either it should be an internal helper trait or there has to be a discussion whether this should become a public API. Hidden unstable APIs are only done to work around limitations of the type system, not as a "well, we might or might not stabilize it" in-betweens.

Additionally this is only used by a specilization. Since the path to stabilization for specialization is unclear we also refrain from adding APIs related to specializations.

Reply:
@the8472 I was trying to implement what you suggested, but I've hit obstacles: the compiler doesn't accept my code (https://github.com/ChayimFriedman2/rust/tree/nonworking-tuple-extend-spec) because:
```
error: cannot specialize on trait `collect::Extend`
   --> library\core\src\iter\traits\collect.rs:512:14
    |
512 |     ExtendB: Extend<B>,
    |              ^^^^^^^^^

error: cannot specialize on trait `collect::Extend`
   --> library\core\src\iter\traits\collect.rs:511:14
    |
511 |     ExtendA: Extend<A>,
    |              ^^^^^^^^^
```
(A false positive, the compiler is not smart enough to understand this is not a bad specialization, but) And:
```
error: specializing impl repeats parameter `T`
    --> library\alloc\src\collections\vec_deque\mod.rs:2854:1
     |
2854 | impl<T, A: Allocator> ExtendUnchecked<T> for VecDeque<T, A> {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
```
I'm afraid today's specialization cannot express that (can you find a way?). So, considering that, and assuming we'll mark the method as perma-unstable, do you still believe we shouldn't do it?

Reply:
I have not been able to find a better approach either. `min_specialization` is quite limiting. Can you leave a (non-doc) comment on the method that it's an internal method and is only on the public trait due to limitations of specialization?

And add a `where Self: Sized` bound so it won't be added to Extend vtables, they have no use for it.